### DISCUSSION TITLE
Question: Advice on testing

### DISCUSSION THREAD

**Discussion by scott-david-walker** [RageAgainstThePixel/OpenAI-DotNet#243] [Q&A]
As the title states, can anybody give some advice about mocking/faking this project? 
Looking at the tests inside the project, they actually test against an actual GPT instance and doesn't provide a way around this?

Thanks

**Comment by StephenHodgson** [comment]
There is a proxy csproj you could probably use to mock this a bit easier. I just like to test against the real deal, since the API changes too frequently.

**Reply by scott-david-walker** [reply]
Yeah I've seen the proxy stuff but my presumption from reading the docs is that this was specifically to solve an authentication problem rather than mocking or testing. Problem is that I'd still end up with the same situation where if I have some flow that calls the chat completion endpoint, I can't actually create a ChatResponse object that I can return (due to the immutability)?. Unless I'm missing something obvious?

**Reply by StephenHodgson** [reply]
I actually use a proxy-test csproj to obfuscate my api key in the unit tests! It's great

**Reply by scott-david-walker** [reply]
Yes but you're testing directly against the OpenApi so you have no need to fake out the responses of ChatGPT. I suppose I'm missing either interfaces or a factory to allow creation of these response objects.

**Reply by StephenHodgson** [reply]
idk, I feel like implementing fake responses is a whole lot of extra work for no reason, when you can test against the real thing.

I guess one way is to "capture" live responses and save them for later and use those for subsequent tests.

**Reply by scott-david-walker** [reply]
Except that testing against GPT is including an external dependency where you're not in control of the result. If you're doing end to end testing then that's acceptable but when Unit Testing or integration testing components in your own system it isn't. I don't want my pipeline to fail because GPT failed or timeout etc. To have to test against an actual instance of GPT to run our test cases seems a little bit crazy to be honest. Especially when you deploy 20+ times a day then you could also see costs increase to a silly level.

**Reply by StephenHodgson** [reply]
That's why I suggested doing one capture for initial responses and then "playing" then back for subsequent tests.

**Reply by scott-david-walker** [reply]
Then we're still interacting with a 3rd party service as a part of the tests. We're also then making unit tests dependent upon another unit test (if we read the first and then playback as you suggest) which defeats the object of unit testing?
Your library is really nice, but the lack of testability makes it difficult to integrate. Especially as services out of your control should be mocked and not a part of unit tests

**Reply by scott-david-walker** [reply]
For anybody interested in a work around. You can add your own http client to the OpenAi client and then create your own MessageHandler (parameter to HttpClient constructor). You can then tell the message handler to return a raw json string that matches the object structure of "ChatResponse".

Not ideal but better than calling GPT as a part of your test suite

**Reply by StephenHodgson** [reply]
I mean... at least there is unit tests 😅 

If you'd like to add mock test suite to the project, I'd greatly appreciate it. I barely have time to add new features as it is.

**Reply by scott-david-walker** [reply]
I mean specifically for people integrating your library. The tests that you have for this project are fine, but I don't want to call GPT when I'm testing my code 😊

I saw that somebody already made a pr that added interfaces that would make this testable by implementers but you rejected it because you didn't want the unnecessary interfaces?

**Reply by StephenHodgson** [reply]
Yeah, it added complexity that isn't needed, and I don't like the DI pattern it enforces  in general

**Reply by scott-david-walker** [reply]
Yes, but to be able to mock or fake any class, it would either need to implement an interface or the class would have to have virtual methods. That alongside the immutability makes it nearly impossible for an implementer to fake the call to GPT when testing.  And I'd argue that being able to fake a call to an external service is of vital importance when unit/integration testing. We could have a factory for each of the response classes that allows them to be created for testing purposes, but this would require the setter to be made internal on all of the necessary properties. I'd happily implement this, but I'd need some agreement from you on how we should approach it, as I'd rather not put in the effort just to end up with no solution. Thoughts?

**Reply by StephenHodgson** [reply]
let me think on it. I'm sure there's a way to test against fake proxy service without changing the core project.

Just seems unnecessary to change your whole client library just to enable mock tests. less is more.

**Reply by scott-david-walker** [reply]
Normally I'd agree, but when your client library acts as an interface to an api, I think it becomes more important. I look forward to hearing your thoughts on a possible solution.

**Reply by baramartin** [reply]
Any news on this? I am currently updating from version "6.8.7" to "7.7.6"  and it broke my working unit tests. 
The issues is I have is the following: 
> I created a wraper around your service for the calls I need in order to be able to test my logic to various responses. With the version I used before, I was able to construct some of your response classes, such as 'ChatResponse', 'Choice', etc.. This is no longer possible due to constructors becoming private. I also tried deserializing them from Json, but obviously that didn't work at all either since the setters are private too.

**Reply by StephenHodgson** [reply]
> Any news on this? I am currently updating from version "6.8.7" to "7.7.6" and it broke my working unit tests.

I suppose take baby steps with each release.
I keep the change log pretty well documented, and document all of the breaking changes and typically follow SemVer.

**Reply by ChrisBrookesWork** [reply]
Any update/suggestions on this. I would llike to test my client code that wraps around calls to the OpenAIClient but I am unable to mock anything due to sealed classes, private setters and lack of interfaces. 
I have hidden all of this library from my code behind an interface but the return type is still of type ChatResponse which i an unable to instantiate in any meaningfull form.

**Reply by scott-david-walker** [reply]
@ChrisBrookesWork - The only way I've found is to override the HttpMessageHandler as a part of the HttpClient. Not ideal but it's ok as a workaround

**Reply by StephenHodgson** [reply]
Literally ever object can be deserialized from json, so just mock your data, deserialize it, and test it against that.

**Reply by StephenHodgson** [reply]
I'm considering a way to mock/save all the test responses to effectively "rerun" them or mock them against specific uses
